[TIMOB-25032] iOS: Resolve asset-catalog hashes back to real names, fail build when detecting duplicates

Sajoha opened the pull request:
JIRA: https://jira.appcelerator.org/browse/TIMOB-25032

Removed hashing from file names, first in the build step, so assets are stored under their real names, with checks in place to make sure duplicates are not used. And also in the module code to ensure de-hashing isn't attempted

Sajoha committed:
[TIMOB-25032] iOS: Resolve asset-catalog hashes back to real names, fail build when detecting duplicates

build commented:
<!--
  2 failure:  :disappointed_rel..., :microscope: Ther...
  0 warning: 
  2 messages
  
  DangerID: danger-id-default;
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  :disappointed_relieved: `npm test` failed. See below for details.
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  :microscope: There are library changes, but no changes to the unit tests. That's OK as long as you're refactoring existing code, but will require an admin to merge this PR. Please see <a href="https://github.com/SamJHaig/titanium_mobile/blob/TIMOB-25032/README.md#unit-tests">README.md#unit-tests</a> for docs on unit testing.
  </td>
    </tr>
  </tbody>
</table>



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  :thumbsup: Hey!, You deleted more code than you added. That's awesome!
  </td>
    </tr>
  
<tr>
      <td>:book:</td>
      <td>

  ```
> titanium-mobile@7.4.0 test /Users/build/jenkins/workspace/dk_titanium_mobile_PR-10239-HVBHSHPIOGQY2U3OG774DOXTQD3O4R2JSE2EQQEJHQQKVETDTOTQ
> grunt

Running "appcJs:src:lintOnly" (appcJs) task

Running "eslint:src" (eslint) task

/Users/build/jenkins/workspace/dk_titanium_mobile_PR-10239-HVBHSHPIOGQY2U3OG774DOXTQD3O4R2JSE2EQQEJHQQKVETDTOTQ/tests/Resources/ti.locale.addontest.js
  8:15  warning  'L' is defined but never used  no-unused-vars

✖ 1 problem (0 errors, 1 warning)


Running "checkFormat:ios" (checkFormat) task
Fatal error: Error: Formatting incorrect on "iphone/Classes/FilesystemModule.m", proposed changes: <?xml version='1.0'?>
<replacements xml:space='preserve' incomplete_format='false'>
<replacement offset='5894' length='10'>&#10;&#10;    </replacement>
<replacement offset='5942' length='10'>&#10;&#10;    </replacement>
</replacements>

Error: Formatting incorrect on "iphone/Classes/ImageLoader.m", proposed changes: <?xml version='1.0'?>
<replacements xml:space='preserve' incomplete_format='false'>
<replacement offset='13166' length='14'>&#10;&#10;      </replacement>
<replacement offset='13230' length='14'>&#10;&#10;      </replacement>
<replacement offset='13288' length='14'>&#10;&#10;      </replacement>
</replacements>

Error: Formatting incorrect on "iphone/Classes/TiUIImageView.m", proposed changes: <?xml version='1.0'?>
<replacements xml:space='preserve' incomplete_format='false'>
<replacement offset='14825' length='14'>&#10;&#10;      </replacement>
<replacement offset='15215' length='14'>&#10;&#10;      </replacement>
<replacement offset='15279' length='14'>&#10;&#10;      </replacement>
<replacement offset='15331' length='14'>&#10;&#10;      </replacement>
<replacement offset='15434' length='14'>&#10;&#10;      </replacement>
</replacements>

npm ERR! Test failed.  See above for more details.

```
  </td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

hansemannn commented on iphone/Classes/FilesystemModule.m:
```diff
@@ -196,20 +196,10 @@ - (id)getAsset:(id)args
-        RELEASE_TO_NIL(sha)
-      }
-    }
+    NSString *fileName = [imageArg lastPathComponent];
```
I think you can reuse the existing `imageArg` property:
```objc
imageArg = [imageArg lastPathComponent];
```
and align it to line 198 to be right after the other string manipulations. Same for other occurences of this code in the core.

hansemannn commented on iphone/cli/commands/_build.js:
```diff
@@ -5706,15 +5706,34 @@ iOSBuilder.prototype.copyResources = function copyResources(next) {
+			const checked = new Set();
+			const duplicates = new Set();
+
+			Object.keys(imageAssets).forEach(function (assetName) {
```
Maybe some ES6 here?
```js
Object.keys(imageAssets).forEach(assetName => {
```

hansemannn commented on iphone/cli/commands/_build.js:
```diff
@@ -5706,15 +5706,34 @@ iOSBuilder.prototype.copyResources = function copyResources(next) {
+			});
+
+			if (duplicates.size > 0) {
+				duplicates.forEach(function (asset) {
```
=>

hansemannn commented on iphone/cli/commands/_build.js:
```diff
@@ -5706,15 +5706,34 @@ iOSBuilder.prototype.copyResources = function copyResources(next) {
+
+			if (duplicates.size > 0) {
+				duplicates.forEach(function (asset) {
+					this.logger.error(__('Image asset name "%s" used multiple time, duplicate names are not permitted', asset));
```
```js
this.logger.error(__('Image asset name "%s" used multiple time, asset names should be unique in Titanium SDK 8.0.0+', asset));
```

hansemannn reviewed (changes requested):
Left a few changes. Also requesting Chris' review for the CLI part of this.

Sajoha committed:
Linting, Fixes, and Test

Sajoha committed:
refactor: Resolved Merge Conflicts

Rebased with master to resolve existing merge conflicts

BREAKING CHANGE: Will throw an error on build if there are duplicated image names in the project

Sajoha committed:
refactor: Bump Warning Message

Changed the warning message from notifying of 8.0.0 to 9.0.0

build commented:
<!--
  1 failure:  :disappointed_rel...
  2 warning:  :mag: Can't find ..., iphone/cli/comman...
  1 messages
  
  DangerID: danger-id-Danger;
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  :disappointed_relieved: `npm test` failed. See below for details.
  </td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  :mag: Can't find junit reports at `./junit.*.xml`, skipping generating JUnit Report.
  </td>
    </tr>
  
<tr>
      <td>:warning:</td>
      <td>

  **iphone/cli/commands/_build.js#L6455** - iphone/cli/commands/_build.js line 6455 – 'out' is defined but never used. Allowed unused args must match /^_.+/. (no-unused-vars)
  </td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  ```
> titanium-mobile@8.1.0 test /Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-10239
> npm run ios-sanity-check && grunt


> titanium-mobile@8.1.0 ios-sanity-check /Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-10239
> ./build/scons check-ios-toplevel

Running "appcJs:src:lintOnly" (appcJs) task

Running "eslint:src" (eslint) task

/Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-10239/iphone/cli/commands/_build.js
  6455:79  warning  'out' is defined but never used. Allowed unused args must match /^_.+/  no-unused-vars

/Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-10239/iphone/cli/hooks/frameworks.js
   61:5   warning  Expected catch() or return                  promise/catch-or-return
   61:34  warning  Avoid calling back inside of a promise      promise/no-callback-in-promise
  428:5   warning  Each then() should return a value or throw  promise/always-return

/Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-10239/tests/Resources/ti.ui.navigationwindow.addontest.js
  11:5  warning  'should' is assigned a value but never used  no-unused-vars

✖ 5 problems (0 errors, 5 warnings)


Running "checkFormat:ios" (checkFormat) task
Fatal error: Error: Formatting incorrect on "iphone/Classes/FilesystemModule.m", proposed changes: <?xml version='1.0'?>
<replacements xml:space='preserve' incomplete_format='false'>
<replacement offset='6810' length='10'>&#10;&#10;    </replacement>
</replacements>

npm ERR! Test failed.  See above for more details.

```
  </td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 59c3af6eb882defe6a90c17957be104f50104463
</p>

cb1kenobi reviewed (commented):
I think this ticket is a bad idea and that we should **NOT** merge it. Please see my comment on the ticket: https://jira.appcelerator.org/browse/TIMOB-25032.

sgtcoolguy committed:
Merge branch 'master' into TIMOB-25032

build commented on tests/Resources/ti.ui.imageview.addontest.js:
```diff
@@ -0,0 +1,52 @@
+ * Please see the LICENSE included with this distribution for details.
+ */
+/* eslint-env mocha */
+/* global Ti */
```
<!--
  1 failure:  tests/Resources/t...
  0 warning: 
  
  
  DangerID: danger-id-Danger;
  File: tests/Resources/ti.ui.imageview.addontest.js;
  Line: 8;
-->
- :no_entry_sign: tests/Resources/ti.ui.imageview.addontest.js line 8 – 'Ti' is already defined as a built-in global variable. (no-redeclare)

vijaysingh-axway commented:
Closing in favor of PR #11437
